data/transactions/logic/eval_test.go

```diff
@@ -5117,3 +5129,399 @@ func TestOpBase64Decode(t *testing.T) {
 	args := b64TestDecodeAssembleWithArgs(t)
 	b64TestDecodeEval(t, args)
 }
+
+func TestHasDuplicateKeys(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	testCases := []struct {
+		text []byte
+	}{
+		{
+			text: []byte(`{"key0": "1","key0": "2", "key1":1}`),
+		},
+		{
+			text: []byte(`{"key0": "1","key1": [1], "key0":{"key2": "a"}}`),
+		},
+	}
+	for _, s := range testCases {
+		hasDuplicates, _, err := hasDuplicateKeys(s.text)
+		require.Nil(t, err)
+		require.True(t, hasDuplicates)
+	}
+
+	noDuplicates := []struct {
+		text []byte
+	}{
+		{
+			text: []byte(`{"key0": "1","key1": "2", "key2":3}`),
+		},
+		{
+			text: []byte(`{"key0": "1","key1": [{"key0":1,"key0":2},{"key0":1,"key0":2}], "key2":{"key5": "a","key5": "b"}}`),
+		},
+	}
+	for _, s := range noDuplicates {
+		hasDuplicates, _, err := hasDuplicateKeys(s.text)
+		require.Nil(t, err)
+		require.False(t, hasDuplicates)
+	}
+}
+
+func TestOpJSONRef(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	proto := defaultEvalProtoWithVersion(LogicVersion)
+	txn := transactions.SignedTxn{
+		Txn: transactions.Transaction{
+			Type: protocol.ApplicationCallTx,
+		},
+	}
+	ledger := logictest.MakeLedger(nil)
+	ledger.NewApp(txn.Txn.Receiver, 0, basics.AppParams{})
+	sb := strings.Builder{}
+	ep := defaultEvalParams(&sb, &txn)
+	ep.Proto = &proto
+	ep.Ledger = ledger
+	testCases := []struct {
+		source             string
+		previousVersErrors []expect
+	}{
+		{
+			source: `byte  "{\"key0\": 0,\"key1\": \"algo\",\"key2\":{\"key3\": \"teal\", \"key4\":3}, \"key5\": 18446744073709551615 }"; 
+			byte "key0"; 
+			json_ref JSONUint64; 
+			int 0; 
+			==`,
+			previousVersErrors: []expect{{5, "unknown opcode: json_ref"}, {9, "== arg 0 wanted type uint64 got []byte"}},
```

Comment:
@shiqizng - I ended adding expected errors for when attempting to run on old TEAL versions